marker_api/src/sem/item.rs

```diff
@@ -0,0 +1,160 @@
+use std::marker::PhantomData;
+
+use crate::common::ItemId;
+
+/// The declared visibility of an item or field.
+///
+/// ```
+/// // An item without a visibility
+/// fn moon() {}
+///
+/// // A public item
+/// pub fn sun() {}
+///
+/// // An item with a restricted scope
+/// pub(crate) fn star() {}
+///
+/// pub trait Planet {
+///     // An item without a visibility. But it still has the semantic visibility
+///     // of `pub` as this is inside a trait declaration.
+///     fn mass();
+/// }
+/// ```
+#[repr(C)]
+#[derive(Debug)]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
+pub struct Visibility<'ast> {
+    #[cfg_attr(feature = "driver-api", builder(setter(skip), default))]
+    _lifetime: PhantomData<&'ast ()>,
+    kind: VisibilityKind,
+}
+
+impl<'ast> Visibility<'ast> {
+    /// Returns `true` if the item is declared as public, without any restrictions.
+    ///
+    /// ```
+    /// // This returns `true`
+    /// pub fn unicorn() {}
+    ///
+    /// // This returns `false`, since the visibility is restricted to a specified path.
+    /// pub(crate) fn giraffe() {}
+    ///
+    /// // This returns `false`, since the visibility is not defined
+    /// fn dragon() {}
+    /// ```
+    ///
+    /// See [`Visibility::is_pub_in_path`] to detect pub declarations with a
+    /// defined path.
+    pub fn is_pub(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Public | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns `true` if the item is declared as `pub(in ..)` with a path in brackets
+    /// that defines the scope, where the item is visible.
+    pub fn is_pub_in_path(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Path(_))
+    }
+
+    /// Returns `true` if the visibility is declared as `pub(crate)`. This is a
+    /// special case of the `pub(<path>)` visibility.
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    // Ignore, since the `in crate::example_1` path doesn't work for doc tests
+    /// ```ignore
+    /// // lib.rs
+    ///
+    /// mod example_1 {
+    ///     // Returns `false` since no visibility is declared
+    ///     fn foo() {}
+    ///
+    ///     // Returns `false` since the item is not visible from the root of the crate.
+    ///     pub(in crate::example_1) fn bar() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to the root of the crate.
+    ///     pub(crate) fn baz() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to `super` which is the
+    ///     // root of the crate.
+    ///     pub(super) fn boo() {}
+    /// }
+    ///
+    /// // Returns `false` since the visibility is not restricted
```

Comment:
Technically, this item is also `pub(crate)`. That's a bit unintuitive. Could we identify that the default visibility at the root of the crate as implicitly `pub(crate)` as well? I guess it would require just another `VisibilityKind`

Reply:
Yes we could, and I would actually guess that this is currently wrongly classified as `pub(crate)`.